landoapi/workers/revision_worker.py

```diff
@@ -0,0 +1,364 @@
+    running.
+    """
+
+    def loop(self):
+        """Run the event loop for the revision worker."""
+        self.throttle()
+        mark_stale_revisions()
+        discover_revisions()
+
+
+class Processor(RevisionWorker):
+    """A worker that pre-processes revisions.
+
+    This worker attempts to import each patch and its predecessors, and detects any
+    issues that come up during the import.
+
+    NOTE: This worker supports scaling and can run independently of other workers.
+    """
+
+    def loop(self):
+        """Run the event loop for the revision worker."""
+        self.throttle()
+
+        # Fetch revisions that require pre-processing.
+        with db.session.begin_nested():
+            Revision.lock_table()
+            revisions = Revision.query.filter(
+                Revision.status.in_([RS.WAITING, RS.STALE])
+            ).limit(self.capacity)
+
+            picked_up = [r.id for r in revisions]
+
+            # Mark revisions as picked up so other workers don't pick them up.
+            Revision.query.filter(Revision.id.in_(picked_up)).update(
+                {Revision.status: RS.PICKED_UP}
+            )
+
+        db.session.commit()
+
+        revisions = Revision.query.filter(Revision.id.in_(picked_up))
+
+        # NOTE: The revisions will be processed according to their dependencies
+        # at the time of fetching. If dependencies change, they will be
+        # re-processed on the next iteration. This has the effect of processing
+        # revisions as they become available, if, for example, a large stack is
+        # being uploaded.
+
+        logger.info(f"Found {revisions.all()} to process.")
+        for revision in revisions:
+            errors = []
+            logger.info(f"Running checks on revision {revision}")
+
+            revision.status = RS.CHECKING
+            db.session.commit()
+
+            try:
+                errors = self.process(revision)
+            except Exception as e:
+                logger.info(f"Exception encountered while processing {revision}")
+                revision.status = RS.PROBLEM
+                revision.update_data(error="".join(e.args))
+                logger.exception(e)
+                db.session.commit()
+                continue
+
+            if errors:
+                logger.info(f"Errors detected on revision {revision}")
+                revision.status = RS.PROBLEM
+                revision.update_data(error="".join(errors))
+            else:
+                revision.status = RS.READY
+                logger.info(f"No problems detected on revision {revision}")
+            db.session.commit()
+
+    def _process_patch(self, revision: Revision, hgrepo: HgRepo) -> list[str]:
+        """Run through all predecessors before applying revision patch."""
+        errors = []
+        for r in revision.predecessors + [revision]:
+            try:
+                hgrepo.apply_patch(io.BytesIO(r.patch_bytes))
+            except Exception as e:
```

Comment:
See [this comment](https://github.com/mozilla-conduit/lando-api/pull/224#discussion_r1031795838).